tgui/packages/tgui/interfaces/NtosMessenger.tsx

```diff
@@ -0,0 +1,469 @@
+  }
+
+  handleMessageInput(_: any, val: string): void {
+    this.setState({ msg: val });
+  }
+
+  render(): InfernoNode | undefined {
+    const { act } = useBackend<NtosMessengerData>(this.context);
+    const { recp, onReturn, msgs } = this.props;
+    const { msg, canSend } = this.state;
+
+    let lastMsgRef = '';
+    let filteredMessages: JSX.Element[] = [];
+
+    for (let index = 0; index < msgs.length; index++) {
+      let message = msgs[index];
+
+      const isSwitch = lastMsgRef !== message.sender;
+      lastMsgRef = message.sender;
+
+      filteredMessages.push(
+        <Stack.Item key={index} mt={isSwitch ? 2 : 0.5}>
+          <ChatMessage
+            isSelf={message.outgoing}
+            msg={message.contents}
+            everyone={!!message.everyone}
+            photoPath={message.photo_path}
+          />
+        </Stack.Item>
+      );
+    }
+
+    return (
+      <Stack vertical fill>
+        <Stack.Item>
+          <Section fill>
+            <Button icon="arrow-left" content="Back" onClick={onReturn} />
+            <Button.Confirm
+              icon="trash-can"
+              content="Delete chat"
+              onClick={() => act('PDA_clearMessages', { ref: recp.ref })}
+            />
+          </Section>
+        </Stack.Item>
+
+        <Stack.Item grow={1}>
+          <Section
+            scrollable
+            fill
+            fitted
+            title={`${recp.name} (${recp.job})`}
+            scrollableRef={this.scrollRef}>
+            <Stack vertical fill className="NtosMessenger__ChatLog">
+              <Stack.Item textAlign="center" fontSize={1}>
+                This is the beginning of your chat with {recp.name}.
+              </Stack.Item>
+              <Stack.Divider />
+              {filteredMessages}
+            </Stack>
+          </Section>
+        </Stack.Item>
+
+        <Stack.Item>
+          <Section>
+            <Stack fill>
+              <Stack.Item grow={1}>
+                <Input
+                  placeholder={`Send message to ${recp.name}...`}
+                  fluid
+                  autofocus
+                  justify
+                  id="input"
+                  value={msg}
+                  maxLength={1024}
+                  onInput={this.handleMessageInput}
+                  onEnter={this.handleSendMessage}
+                />
+              </Stack.Item>
+              <Stack.Item>
+                <Button
+                  icon="arrow-right"
+                  onClick={this.handleSendMessage}
+                  disabled={!canSend}
+                />
+              </Stack.Item>
+            </Stack>
+          </Section>
+        </Stack.Item>
+      </Stack>
+    );
+  }
+}
+
+const SendToAllModal = (_: any, context: any) => {
+  const { data, act } = useBackend<NtosMessengerData>(context);
+  const { on_spam_cooldown } = data;
+
+  const [msg, setMsg] = useLocalState<string>(context, 'everyoneMessage', '');
```

Comment:
i could be wrong about this with our useLocalState, but this should be inferred unless you're using something that needs a union. meaning, untyped, `useLocalState(context...` with no `<string>`, as we should just infer the type as string